axum-login/src/middleware.rs

```diff
@@ -179,7 +202,14 @@ macro_rules! predicate_required {
                         original_uri
                     ) {
                         Ok(login_url) => {
-                            Redirect::temporary(&login_url.to_string()).into_response()
+                            if $redirect_status_code == $crate::axum::http::StatusCode::TEMPORARY_REDIRECT {
```

Comment:
I think this is a sign our design might be moving in the wrong direction.

How about instead of a static set of HTTP codes, we accept a response producer that takes `login_url` as an argument and then returns `impl IntoResponse`? This way we don't need to worry about the status codes explicitly and consumers are able to do whatever they want when the login fails (including return 401 Unauthorized).

Reply:
I also had same concern while writing that line.
Can you give an example for response producer? I'm not sure how to implement it

Reply:
An expression that takes the login URL, e.g. `$failed_predicate_callback(login_url.to_string())`, might work.

Probably a similar treatment is applicable to the error branch as well and more generally, it might be good to take a step back and consider the flexibility of the API. (It's going to get messy quickly, when `redirect_field` and lack thereof are considered as branches are set up to account for a callback.)

At the end of the day, `predicate_required` already has the most flexibility when using a predicate and alternative (its first match).

Reply:
I've updated the code to include failed_predicate_callback option, but it accepts `original_url` so user can handle it as per their requirements.

With this implementation "predicate and alternative" in `predicate_required` is bit redundant as its case can be covered using a failed_predicate_callback, but I kept it as is.